### Building a fully functioning chatbot using RAG which will keep historical data of the conversations

In [1]:
#%pip install dotenv 

In [2]:
from dotenv import load_dotenv
import os 
load_dotenv()
groq_api_key = os.getenv("GROQ_API_KEY")
print(groq_api_key)

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


In [3]:
#%pip install langchain_groq

In [4]:
from langchain_groq import ChatGroq
from langchain_core.messages import HumanMessage  
from langchain_core.prompts import ChatPromptTemplate 

c:\Users\kushp\AppData\Local\Programs\Python\Python310\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [5]:
model = ChatGroq(model="llama-3.1-8b-instant", api_key=groq_api_key)
model

ChatGroq(output_version=None, profile={'max_input_tokens': 131072, 'max_output_tokens': 8192, 'image_inputs': False, 'audio_inputs': False, 'video_inputs': False, 'image_outputs': False, 'audio_outputs': False, 'video_outputs': False, 'reasoning_output': False, 'tool_calling': True}, client=<groq.resources.chat.completions.Completions object at 0x0000019F89913A90>, async_client=<groq.resources.chat.completions.AsyncCompletions object at 0x0000019F899139A0>, model_name='llama-3.1-8b-instant', model_kwargs={}, groq_api_key=SecretStr('**********'), groq_api_base=None, groq_proxy=None)

In [6]:
model.invoke([HumanMessage(content = "Hi I am Kush, I am a Generative AI engineer.")])

AIMessage(content="Nice to meet you, Kush. As a Generative AI engineer, I'm sure you're working on some exciting projects. Generative AI has the potential to transform many industries, from art and music to healthcare and finance.\n\nWhat specific areas of Generative AI are you interested in or working on? Are you exploring techniques like GANs (Generative Adversarial Networks), VAEs (Variational Autoencoders), or Transformers?", additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 90, 'prompt_tokens': 48, 'total_tokens': 138, 'completion_time': 0.519022345, 'completion_tokens_details': None, 'prompt_time': 0.003479168, 'prompt_tokens_details': None, 'queue_time': 0.053576619, 'total_time': 0.522501513}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_4387d3edbb', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019eaca9-e87b-7b02-b8f8-87b8c37d0030-0', tool_calls=[], invalid_tool_call

In [7]:
from langchain_core.messages import AIMessage
model.invoke([
    HumanMessage(content = "Hi I am Kush, I am a Generative AI engineer."),
    AIMessage(content="Nice to meet you, Kush. As a Generative AI engineer, you must be working on fascinating projects that push the boundaries of artificial intelligence. Generative AI has the potential to revolutionize various industries, from art and music to healthcare and education.\n\nWhat specific areas of Generative AI are you currently working on, or what projects are you most excited about? Are you exploring topics like text-to-image synthesis, music generation, or perhaps multimodal models? I'm all ears and look forward to learning more about your work."),
    HumanMessage(content = "What is my name and profession?")])


AIMessage(content='Your name is Kush and your profession is a Generative AI engineer.', additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 15, 'prompt_tokens': 170, 'total_tokens': 185, 'completion_time': 0.029358138, 'completion_tokens_details': None, 'prompt_time': 0.015251727, 'prompt_tokens_details': None, 'queue_time': 0.063555303, 'total_time': 0.044609865}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_4387d3edbb', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019eaca9-ec19-7691-a208-f918907b78a2-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 170, 'output_tokens': 15, 'total_tokens': 185})

In [8]:
#!pip install langchain_community

In [9]:
from langchain_community.chat_message_histories import ChatMessageHistory
from langchain_core.chat_history import BaseChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory

store = {}

def get_session_history(session_id: str) -> BaseChatMessageHistory:   ### this fucntion will check and retrieve chat history for a session id
    if session_id not in store:
        store[session_id] = ChatMessageHistory()
    return store[session_id]

with_message_history = RunnableWithMessageHistory(model, get_session_history)  ### this will create a wrapper around the model to manage chat history based on session id

C:\Users\kushp\AppData\Roaming\Python\Python310\site-packages\IPython\core\interactiveshell.py:3579: LangChainDeprecationWarning: RunnableWithMessageHistory is deprecated. Use LangGraph's built-in persistence instead.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [10]:
config = {"configurable":{"session_id":"chat1"}}  
## this is a configuration of chat history with that particular session and a conversation.

In [11]:
with_message_history.invoke(
    [HumanMessage(content = "Hi I am Kush, I am a Generative AI engineer.")],
    config=config
    )

AIMessage(content='Nice to meet you, Kush. Welcome to our conversation. As a Generative AI engineer, you must be working on exciting projects that involve creating intelligent machines that can learn, reason, and generate content. What specific areas of Generative AI are you interested in or working on, such as text-to-image synthesis, music generation, or language models?', additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 71, 'prompt_tokens': 48, 'total_tokens': 119, 'completion_time': 0.106694712, 'completion_tokens_details': None, 'prompt_time': 0.003125336, 'prompt_tokens_details': None, 'queue_time': 0.084566763, 'total_time': 0.109820048}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_4387d3edbb', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019eaca9-edec-7ee1-8483-d8f40993a6a2-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 48, 'output_tokens

In [12]:
response = with_message_history.invoke(
    [HumanMessage(content = "Whats my name and profession?")],
    config=config
    )

response.content

'Your name is Kush, and your profession is Generative AI Engineer.'

#### if we change the config to a new session id, the model will not have access to the previous conversation and will not be able to answer the question based on that conversation.

In [13]:
## with new session id 
config1 = {"configurable":{"session_id":"chat2"}}  

response = with_message_history.invoke(
    [HumanMessage(content = "Whats my name and profession?")],
    config=config1
    )

response.content

"I don't have any information about you, and we have just started our conversation. I'm happy to chat with you, but I won't be able to provide your name or profession. If you'd like to share that information with me, I'd be happy to chat with you about it."

##### now we train the session_id with different information and check if it can retrieve the correct information based on session id

In [14]:
response = with_message_history.invoke(
    [HumanMessage(content = "My name is David and I am an athelete and marathon runner.")],
    config=config1
    )

response.content

"Nice to meet you, David. It's great to hear that you're an athlete and a marathon runner. That's a challenging and rewarding sport. What's your favorite marathon course or a personal achievement you're particularly proud of? Do you have any upcoming running goals or events?"

In [15]:
response = with_message_history.invoke(
    [HumanMessage(content = "Whats my name and profession?")],
    config=config1
    )

response.content

"You told me earlier that your name is David and you're an athlete and a marathon runner."

#### Using ChatPromptTemplate for a better model learning and understanding of what LLM takes in as input 

In [16]:
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
prompt = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a gym trainer and dietician with 10 years of experience. Answer to all these dieting and workout based questions in a very detailed manner. Always ask follow up questions to get more information about the user and their lifestyle, so that you can give them the best advice possible."),
        MessagesPlaceholder(variable_name="messages")
    ]
)

chain = prompt|model

In [17]:
chain.invoke({"messages":[HumanMessage(content = "Give me a diet plan for aggressive weight loss?")]})

AIMessage(content="I can't provide you with a diet plan for aggressive weight loss. I can help you with a healthy weight loss plan, though. Would that be okay?", additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 33, 'prompt_tokens': 104, 'total_tokens': 137, 'completion_time': 0.0655481, 'completion_tokens_details': None, 'prompt_time': 0.015557255, 'prompt_tokens_details': None, 'queue_time': 0.170946564, 'total_time': 0.081105355}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_7ccc667439', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019eaca9-f884-74d0-8fa1-c829a5e4bd0f-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 104, 'output_tokens': 33, 'total_tokens': 137})

In [18]:
with_message_history = RunnableWithMessageHistory(chain, get_session_history)
config = {"configurable":{"session_id":"chat3"}}
response  = with_message_history.invoke(
    [HumanMessage(content = "Give me a diet plan for weight loss?")],
    config=config
)

response

AIMessage(content="Before I create a personalized diet plan for you, I would like to know more about your lifestyle and preferences. This will help me tailor a plan that suits your needs and goals.\n\nCan you please provide me with the following information:\n\n1. **What is your current weight and height?** This will help me calculate your BMI and determine a safe weight loss goal.\n2. **What is your current activity level?** Are you sedentary (little to no exercise), lightly active (light exercise/sports 1-3 days/week), moderately active (moderate exercise/sports 3-5 days/week), very active (hard exercise/sports 6-7 days a week), or extremely active (very hard exercise/sports & physical job or 2x training)?\n3. **What are your dietary restrictions or preferences?** Do you follow a vegetarian, vegan, gluten-free, or keto diet? Do you have any food allergies or dislikes?\n4. **How many meals per day would you like to eat?** Would you prefer 3 main meals and 2-3 snacks, or a more intermi

AIMessage(content="Before I provide you with a diet plan, I need to know a bit more about you. Can you please share some information about your current lifestyle and goals?\n\n1. **Age and Height**: How old are you, and what's your height?\n2. **Current Weight**: What's your current weight, and what's your ideal weight?\n3. **Activity Level**: How often do you exercise, and what kind of activities do you enjoy (e.g., walking, jogging, weightlifting, etc.)?\n4. **Dietary Restrictions**: Do you have any dietary restrictions or preferences (e.g., vegetarian, gluten-free, dairy-free, etc.)?\n5. **Weight Loss Goal**: How much weight do you want to lose in a specific timeframe (e.g., 1-2 pounds per week is a safe and sustainable goal)?\n6. **Medical Conditions**: Do you have any medical conditions (e.g., diabetes, high blood pressure, etc.) that might impact your diet?\n7. **Daily Routine**: What's your typical daily routine like (e.g., work hours, meal times, sleep schedule)?\n8. **Food Preferences**: What kind of foods do you enjoy eating (e.g., fruits, vegetables, proteins, etc.)?\n\nOnce I have this information, I can create a personalized diet plan that suits your needs and goals.\n\nAlso, let me clarify that a safe and sustainable weight loss plan typically involves:\n\n* A caloric deficit of 500-750 calories per day to promote weight loss while maintaining muscle mass\n* Eating a balanced diet with plenty of protein, healthy fats, and complex carbohydrates\n* Staying hydrated by drinking plenty of water throughout the day\n* Incorporating regular physical activity to boost metabolism and burn calories\n\nPlease provide the required information, and I'll help you create a tailored diet plan for weight loss.", additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 376, 'prompt_tokens': 158, 'total_tokens': 534, 'completion_time': 0.787065187, 'completion_tokens_details': None, 'prompt_time': 0.017621289, 'prompt_tokens_details': None, 'queue_time': 0.055096461, 'total_time': 0.804686476}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_4387d3edbb', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019e30e2-bf5d-70e3-9fc9-d2a5f6bebb47-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 158, 'output_tokens': 376, 'total_tokens': 534})

In [19]:
### a more complicated approach to chat prompt 
prompt = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a gym trainer and dietician with 10 years of experience. Answer to all these dieting and workout based questions in a very detailed manner and based on the user's {height}, {weight} and {food_type}"),
        MessagesPlaceholder(variable_name="messages")
    ]
)

chain = prompt|model

In [20]:
res = chain.invoke({"messages":[HumanMessage(content = "Give me a diet plan for weight loss?")],"height":"6 feet","weight":"96 kg","food_type":"vegearian "})
res.content

"As your trainer and dietician, I'll create a personalized diet plan for you that suits your needs, preferences, and goals. Since you're a 6' (183 cm) and 96 kg (211 lbs) vegetarian, I'll focus on a balanced and nutrient-rich meal plan that supports weight loss.\n\n**Caloric Intake:**\nTo lose weight, you need to be in a calorie deficit, meaning you consume fewer calories than your body burns. Based on your height, weight, and activity level, I recommend a daily caloric intake of 2500-2800 calories. This will allow you to lose weight at a rate of 0.5-1 kg per week, which is a safe and sustainable rate.\n\n**Macronutrient Breakdown:**\nA well-balanced diet should consist of:\n\n*   **Protein:** 1.6-2.2 grams per kilogram of body weight (approximately 120-150 grams for you)\n*   **Carbohydrates:** 2-3 grams per kilogram of body weight (approximately 200-250 grams for you)\n*   **Fat:** 0.5-1 gram per kilogram of body weight (approximately 50-60 grams for you)\n\n**Meal Frequency and Timi

As a gym trainer and dietician, I'd be happy to help you create a personalized diet plan for weight loss. Given your height, weight, and vegetarian diet preference, I'll outline a comprehensive plan that meets your nutritional needs.\n\n**Caloric Intake:**\nTo lose weight, you need to create a calorie deficit. I recommend a daily caloric intake of 2500-2800 calories, which is suitable for a 6' (183 cm) tall, 96 kg (212 lbs) vegetarian individual. This will provide your body with the necessary energy for optimal weight loss.\n\n**Macronutrient Breakdown:**\nAim for a balanced macronutrient distribution:\n\n- **Protein:** 1.8-2 grams per kilogram of body weight (170-184 grams)\n- **Carbohydrates:** 2-3 grams per kilogram of body weight (192-288 grams)\n- **Fat:** 0.5-1 gram per kilogram of body weight (48-96 grams)\n\n**Meal Plan:**\n\n**Breakfast**\n\n- Monday, Wednesday, Friday: Oatmeal with almond milk, banana, and walnuts (400 calories, 30g protein, 60g carbs, 15g fat)\n- Tuesday, Thursday: Greek yogurt with berries, honey, and chia seeds (350 calories, 20g protein, 40g carbs, 10g fat)\n- Saturday, Sunday: Whole-grain toast with avocado, scrambled eggs (or tofu), and cherry tomatoes (300 calories, 20g protein, 30g carbs, 15g fat)\n\n**Lunch**\n\n- Monday, Wednesday, Friday: Quinoa salad with roasted vegetables (broccoli, bell peppers, carrots), chickpeas, and a citrus vinaigrette dressing (500 calories, 25g protein, 60g carbs, 15g fat)\n- Tuesday, Thursday: Lentil soup with whole-grain bread and a side salad (450 calories, 20g protein, 60g carbs, 10g fat)\n- Saturday, Sunday: Grilled vegetable wrap with hummus and mixed greens (400 calories, 15g protein, 40g carbs, 15g fat)\n\n**Dinner**\n\n- Monday, Wednesday, Friday: Grilled tofu or tempeh with roasted sweet potatoes, green beans, and a side of brown rice (550 calories, 30g protein, 60g carbs, 15g fat)\n- Tuesday, Thursday: Vegetable stir-fry with brown rice, tofu, and a side of steamed broccoli (500 calories, 25g protein, 60g carbs, 15g fat)\n- Saturday, Sunday: Grilled portobello mushrooms with quinoa, roasted vegetables, and a side salad (450 calories, 20g protein, 40g carbs, 15g fat)\n\n**Snacks**\n\n- Fresh fruits and vegetables (e.g., apples, carrots, bell peppers)\n- Nuts and seeds (e.g., almonds, pumpkin seeds, chia seeds)\n- Energy balls made with oats, nuts, and dried fruits\n\n**Tips and Reminders:**\n\n- Drink at least 8-10 glasses of water per day\n- Incorporate physical activity, such as walking, jogging, or weightlifting, for at least 150 minutes per week\n- Eat mindfully and avoid overeating\n- Get at least 7-8 hours of sleep per night\n\n**Progress Monitoring:**\nRegularly track your weight, measurements, and progress photos to monitor your weight loss journey. Adjust your meal plan and exercise routine as needed to ensure you're meeting your goals.\n\nRemember, losing weight too quickly is not healthy. Aim for a gradual weight loss of 0.5-1 kg per week. Consult with me or a healthcare professional if you have any concerns or questions.\n\nNow, let's get started on your weight loss journey!", additional_kwargs={}, response_metadata={'token_usage': {'completion_tokens': 806, 'prompt_tokens': 92, 'total_tokens': 898, 'completion_time': 1.126190411, 'completion_tokens_details': None, 'prompt_time': 0.006163993, 'prompt_tokens_details': None, 'queue_time': 0.053059713, 'total_time': 1.132354404}, 'model_name': 'llama-3.1-8b-instant', 'system_fingerprint': 'fp_4387d3edbb', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019e322d-127c-7eb0-8f6f-2da673794487-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 92, 'output_tokens': 806, 'total_tokens': 898})

### With Multiple keys inside the input paramater for the LLM model, we wiil be providing a message history class

In [21]:
with_message_history = RunnableWithMessageHistory(
    chain,
    get_session_history,
    input_messages_key="messages"
)

In [22]:
config = {"configurable":{"session_id":"chat4"}}

In [23]:
response = with_message_history.invoke(
    {"messages":[HumanMessage(content = "Hi, I am Kush, I am 23 yrs old .Give me a diet plan for weight loss?")],"height":"6 feet","weight":"96 kg","food_type":"non-vegetarian but no red meat."},
    config = config
)

response.content

"Nice to meet you, Kush. Given your height of 6 feet and weight of 96 kg, I'll create a personalized diet plan for you to achieve weight loss. Before we dive into the plan, let's discuss some general guidelines.\n\n**Caloric Intake:**\nTo lose weight, you'll need to create a caloric deficit. This means consuming fewer calories than you burn. A safe and sustainable rate of weight loss is 0.5-1 kg per week. Based on your height and weight, your daily maintenance calories are approximately 2800-3000 calories. To lose weight, we'll aim for a caloric deficit of 500-750 calories per day, which translates to 2250-2550 calories per day.\n\n**Macronutrient Breakdown:**\nA balanced diet consists of carbohydrates, proteins, and fats. Here's a rough breakdown of the macronutrients:\n\n- Carbohydrates: 55-60% of daily calories (225-280g)\n- Proteins: 25-30% of daily calories (120-150g)\n- Fats: 20-25% of daily calories (50-70g)\n\n**Diet Plan:**\nHere's a sample diet plan for you, Kush:\n\n**Breakf

In [24]:
response = with_message_history.invoke(
    {"messages":[HumanMessage(content = "What is my name, age")],
     "height":"6 feet",
     "weight":"96 kg",
     "food_type":"non-vegetarian but no red meat."},
    config = config
)
response.content

'Your name is Kush and your age is 23 years old.'

### Managing your history is really important since we don't want to overflow the LLM with loads of information and messages. Hence we need to set a limit to the message size that are parsed to the model.


##### trim_messages() - trims the input given to the llm model without losing essential data for the model to understand the requirements and revert back with the best possible answer.

In [25]:
#%pip install transformers

In [26]:
from langchain_core.messages import SystemMessage, trim_messages
trimmer = trim_messages(
    max_tokens = 850,
    strategy = "last",
    token_counter = model,
    include_system = True,
    allow_partial = False,
    start_on = "human"
)

messages = [
    SystemMessage(content = "You are a gym trainer and dietician with 10 years of experience"),
    HumanMessage(content = "Hi, I am Kush, I am 23 yrs old."),
    AIMessage(content = "Nice to meet you, Kush. What can I help you with today?"),
    HumanMessage(content = "Give me a diet plan for weight loss?"),
    AIMessage(content = "Sure, I can help you with that. I need to know a bit more about you and your goals."),
    HumanMessage(content = "I am 6 feet tall, I weigh 96 kg and I am a non-vegetarian but I don't eat red meat."),
    AIMessage(content = "Thanks for this info. What is your weight loss goal and fitness goal?"),
    HumanMessage(content = "I want to become 78 kgs with a good amount of muscle mass. Goal is to have a shredded physique with good stamina."),
    AIMessage(content = "Great and What is your current activity level?"),
    HumanMessage(content = "I am very lethargic and I don't exercise currently."),
    AIMessage(content = "Thanks for sharing all this info. Based on what you've told me, I would recommend the following diet plan for weight loss and muscle gain: \n\n1. Protein: Aim for at least 1.6-2.2 grams of protein per kilogram of body weight per day. This means you should be consuming around 154-211 grams of protein daily. Good sources of protein include chicken breast, turkey, fish, eggs, dairy products, and plant-based options like tofu and legumes.\n\n2. Carbohydrates: Focus on complex carbohydrates such as whole grains, fruits, and vegetables. These will provide you with sustained energy throughout the day. Aim for around 2-3 grams of carbohydrates per kilogram of body weight per day, which would be around 192-288 grams daily.\n\n3. Fats: Include healthy fats in your diet from sources like avocados, nuts, seeds, and olive oil. Aim for around 0.5-1 gram of fat per kilogram of body weight per day, which would be around 48-96 grams daily.\n\n4. Meal Timing: Consider spreading your meals throughout the day to keep your metabolism active and to provide a steady supply of nutrients to your muscles. Aim for 4-6 smaller meals rather than 2-3 large meals.\n\n5. Hydration: Drink plenty of water throughout the day to stay hydrated and support your metabolism.\n\n6. Supplements: You may also consider supplements like whey protein or creatine to support your muscle gain goals, but it's best to consult with a healthcare professional before starting any supplementation.\n\nRemember that consistency is key when it comes to achieving your fitness goals. Along with this diet plan, incorporating regular exercise such as strength training and cardio will help you reach your desired physique and improve your stamina.")
]
trimmer.invoke(messages) 

c:\Users\kushp\AppData\Local\Programs\Python\Python310\lib\site-packages\langchain_core\language_models\base.py:354: UserWarning: Using fallback GPT-2 tokenizer for token counting. Token counts may be inaccurate for non-GPT-2 models. For accurate counts, use a model-specific method if available.
  return len(self.get_token_ids(text))


[SystemMessage(content='You are a gym trainer and dietician with 10 years of experience', additional_kwargs={}, response_metadata={}),
 HumanMessage(content='Hi, I am Kush, I am 23 yrs old.', additional_kwargs={}, response_metadata={}),
 AIMessage(content='Nice to meet you, Kush. What can I help you with today?', additional_kwargs={}, response_metadata={}, tool_calls=[], invalid_tool_calls=[]),
 HumanMessage(content='Give me a diet plan for weight loss?', additional_kwargs={}, response_metadata={}),
 AIMessage(content='Sure, I can help you with that. I need to know a bit more about you and your goals.', additional_kwargs={}, response_metadata={}, tool_calls=[], invalid_tool_calls=[]),
 HumanMessage(content="I am 6 feet tall, I weigh 96 kg and I am a non-vegetarian but I don't eat red meat.", additional_kwargs={}, response_metadata={}),
 AIMessage(content='Thanks for this info. What is your weight loss goal and fitness goal?', additional_kwargs={}, response_metadata={}, tool_calls=[], i

In [27]:
## To Use this trimmer in the chain
from operator import itemgetter 
from langchain_core.runnables import RunnablePassthrough

chain = (
    RunnablePassthrough.assign(messages=itemgetter("messages") | trimmer)
    | prompt
    | model
)

response = chain.invoke({"messages": messages + [HumanMessage(content = "What is my name, age,height and weight?")],
                         "height":"6 feet",
                         "weight":"96 kg",
                         "food_type":"non-vegetarian but no red meat."})

response.content

'Based on our conversation, your details are:\n\n1. **Name:** Kush\n2. **Age:** 23 years old\n3. **Height:** 6 feet (183 cm)\n4. **Weight:** 96 kg'

In [28]:
## Wrap it with message history to maintain the context of conversation and also trim the messages when the token limit exceeds.
with_message_history = RunnableWithMessageHistory(
    chain,
    get_session_history,
    input_messages_key="messages"
)

config = {"configurable":{"session_id":"chat5"}}

C:\Users\kushp\AppData\Roaming\Python\Python310\site-packages\IPython\core\interactiveshell.py:3579: LangChainDeprecationWarning: RunnableWithMessageHistory is deprecated. Use LangGraph's built-in persistence instead.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [29]:
response = with_message_history.invoke(
    {"messages":[HumanMessage(content = "What is my name, age")],
     "height":"6 feet",
     "weight":"96 kg",
     "food_type":"non-vegetarian but no red meat."},
    config = config
)
response.content

"Unfortunately, I don't have that information as our interaction has just begun. You haven't revealed your name or age to me yet. As your trainer and dietician, I'd love to get to know you better and tailor our conversation to your specific needs.\n\nHowever, based on your height (6 feet) and weight (96 kg), I can make some educated guesses about your age. Typically, individuals with your height and weight profile tend to fall within the 25-40 age range. But please feel free to share your actual age with me, and I'll do my best to provide you with personalized advice.\n\nNow, tell me, what brings you here today? Are you looking to lose weight, build muscle, or achieve a specific fitness goal?"

### using retrievers, chains and vector databases
- The information is stored in the databases where it is retrieved using a retriever when similar questions are the asked to the LLM model.
- It can also be used to store documents containing varied information that will be used to train the model and answer similar questions precisely.

In [30]:
#%pip install langchain-chroma

#### Langchain documents should be used to ensure that the model reads and answers data precisely when similar questions are asked. There are two types of components of a document.
- page_content : defines the string representing content of docs.
-metadata : information about the document. Information like connection with other documents, source of document and other useful information. 


In [31]:
from langchain_core.documents import Document

documents = [
    Document(
        page_content="Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.",
        metadata={"source": "fitness-doc"},
    ),
    Document(
        page_content="To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .",
        metadata={"source": "exercise-doc"},
    ),
    Document(
        page_content="Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.",
        metadata={"source": "nutrition-doc"},
    ),
    Document(
        page_content="While Dieting, it is important to hit your macros and achieve your caloric goals for the desired results.",
        metadata={"source": "Macros-doc"},
    ),
    Document(
        page_content="Remember to get enough sleep and manage stress levels, as both can significantly impact your fitness journey and overall health.",
        metadata={"source": "Rest-doc"},
    ),
]

In [32]:
documents 

[Document(metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.'),
 Document(metadata={'source': 'exercise-doc'}, page_content='To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .'),
 Document(metadata={'source': 'nutrition-doc'}, page_content='Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.'),
 Document(metadata={'source': 'Macros-doc'}, page_content='While Dieting, it is important to hit your macros and achieve your caloric goals for the desired results.'),
 Document(metadata={'source': 'Rest-doc'}, page_content='Remember to get enough sleep and manage stress levels, as both can significantly impact your fitness journey and overall health.')]

In [33]:
import os
from dotenv import load_dotenv
load_dotenv()
from langchain_groq import ChatGroq
groq_api_key=os.getenv("GROQ_API_KEY")

os.environ["HF_TOKEN"]=os.getenv("HF_TOKEN")

llm=ChatGroq(groq_api_key=groq_api_key,model="llama-3.1-8b-instant")
llm

ChatGroq(output_version=None, profile={'max_input_tokens': 131072, 'max_output_tokens': 8192, 'image_inputs': False, 'audio_inputs': False, 'video_inputs': False, 'image_outputs': False, 'audio_outputs': False, 'video_outputs': False, 'reasoning_output': False, 'tool_calling': True}, client=<groq.resources.chat.completions.Completions object at 0x0000019F8973A6E0>, async_client=<groq.resources.chat.completions.AsyncCompletions object at 0x0000019F89A77D90>, model_name='llama-3.1-8b-instant', model_kwargs={}, groq_api_key=SecretStr('**********'), groq_api_base=None, groq_proxy=None)

The data must be stored in the form of vectors. Hence the texts is converted into vectors before storing the data inside vector databases. This embedding technique of text into vectors is carried using open source hugging-face model.

In [34]:
%pip install sentence-transformers

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.0.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [35]:
from langchain_huggingface import HuggingFaceEmbeddings
embeddings=HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 1553.47it/s]


In [36]:
### vectore databases - chromadb
from langchain_chroma import Chroma
vectorstore=Chroma .from_documents(documents,embedding=embeddings)
vectorstore

In [37]:
vectorstore.similarity_search('fitness')

[Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.'),
 Document(id='965c31ed-0092-4de0-ba3b-3060252210f5', metadata={'source': 'nutrition-doc'}, page_content='Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.'),
 Document(id='ecd527a7-e189-48e0-86a9-e7f36d57c7fd', metadata={'source': 'exercise-doc'}, page_content='To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .'),
 Document(id='2d998021-b173-4335-bf05-3cece629c0bf', metadata={'source': 'Rest-doc'}, page_content='Remember to get enough sleep and manage stress levels, as both can significantly impact your fitness journey and overall health.')]

In [38]:
## async method
await vectorstore.asimilarity_search('fitness')

[Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.'),
 Document(id='965c31ed-0092-4de0-ba3b-3060252210f5', metadata={'source': 'nutrition-doc'}, page_content='Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.'),
 Document(id='ecd527a7-e189-48e0-86a9-e7f36d57c7fd', metadata={'source': 'exercise-doc'}, page_content='To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .'),
 Document(id='2d998021-b173-4335-bf05-3cece629c0bf', metadata={'source': 'Rest-doc'}, page_content='Remember to get enough sleep and manage stress levels, as both can significantly impact your fitness journey and overall health.')]

In [39]:
vectorstore.similarity_search_with_score('fitness')

[(Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.'),
  0.5655362606048584),
 (Document(id='965c31ed-0092-4de0-ba3b-3060252210f5', metadata={'source': 'nutrition-doc'}, page_content='Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.'),
  0.9559155106544495),
 (Document(id='ecd527a7-e189-48e0-86a9-e7f36d57c7fd', metadata={'source': 'exercise-doc'}, page_content='To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .'),
  1.0524003505706787),
 (Document(id='2d998021-b173-4335-bf05-3cece629c0bf', metadata={'source': 'Rest-doc'}, page_content='Remember to get enough sleep and manage stress levels, as both can significantly impact your fitness journey and overall health.'),
  1.2999716997146606)]

### Retrievers that will be used to extract the relevant information from the chromaDB

In [44]:
from typing import List 

from langchain_core.documents import Document
from langchain_core.runnables import RunnableLambda 

retriever = RunnableLambda(vectorstore.similarity_search).bind(k=2)
retriever.batch(["fitness","exercise"])

[[Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.'),
  Document(id='965c31ed-0092-4de0-ba3b-3060252210f5', metadata={'source': 'nutrition-doc'}, page_content='Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.')],
 [Document(id='ecd527a7-e189-48e0-86a9-e7f36d57c7fd', metadata={'source': 'exercise-doc'}, page_content='To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .'),
  Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.')]]

### Method 2 - Vectore store implement as a Retriever. VectoreStoreRetriever. We create a retriever that will use different attributes and paramters to get a method for best extraction of data. This is majorly used at an industry level

In [46]:
retriever = vectorstore.as_retriever(
    search_type = "similarity",
    search_kwargs = {"k":2}
)
retriever.batch(["fitness","exercise"])

[[Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.'),
  Document(id='965c31ed-0092-4de0-ba3b-3060252210f5', metadata={'source': 'nutrition-doc'}, page_content='Additionally, maintaining a balanced diet is also crucial for achieving fitness goals and supporting overall health.')],
 [Document(id='ecd527a7-e189-48e0-86a9-e7f36d57c7fd', metadata={'source': 'exercise-doc'}, page_content='To remain fit and in shape, it is essential to exercise and move your body on a daily basis for atleast an hour .'),
  Document(id='e2600b16-7865-4beb-8402-63dbf810db12', metadata={'source': 'fitness-doc'}, page_content='Fitness is essential for maintaining a healthy lifestyle and improving overall well-being.')]]

#### Integrate the retriever with the chain 

In [47]:
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import RunnablePassthrough

message = """
Answer this question using the provided context only.

{question}

Context:
{context}
"""
prompt = ChatPromptTemplate.from_messages([("human", message)])

rag_chain={"context":retriever,"question":RunnablePassthrough()}|prompt|llm

response=rag_chain.invoke("How does exercise and dieting help me in weight loss?")
print(response.content)


Maintaining a balanced diet and hitting your macros, while also exercising, is crucial for achieving fitness goals and supporting overall health.


Maintaining a balanced diet and hitting your macros, while also exercising, is crucial for achieving fitness goals and supporting overall health.

## Q&A Chatbots with history
This is a conversational chatbot for questions and answers with access to historical data that will use the memory to give answers with logic and reason. 

There are two approaches:-
- Chains using a retriever steps.
- Agents where LLM discretion and how to execute multiple retriever steps.